pkgs/fc/neighbour-cache-monitor/neighbour-cache-monitor.py

```diff
@@ -0,0 +1,158 @@
+#!/usr/bin/env python3
+
+import argparse
+import json
+import subprocess
+import sys
+
+CACHE_TYPES = {
+    "ipv4": "arp_cache",
+    "ipv6": "ndisc_cache",
+}
+
+
+def warning(msg):
+    print(f"WARN - {msg}")
+    sys.exit(1)
+
+
+def critical(msg):
+    print(f"CRITICAL - {msg}")
+    sys.exit(2)
+
+
+def ok(msg):
+    print(f"OK - {msg}")
+    sys.exit(0)
+
+
+def read_cache_stats(cache_type):
+    # XXX: lnstat unconditionally returns exit code 1, even on
+    # success.
+    proc = subprocess.run(
+        [
+            "lnstat",
+            "-f",
+            cache_type,
+            "-j",
+            "-c1",
+        ],
+        capture_output=True,
+    )
+    return json.loads(proc.stdout.decode("utf-8"))
+
+
+def read_table_threshold(family):
+    data = subprocess.check_output(
+        ["sysctl", "-n", f"net.{family}.neigh.default.gc_thresh3"]
+    )
+    data = data.strip()
+    return int(data)
+
+
+def telegraf(args):
+    # pass through lnstat output to telegraf verbatim, annotated with
+    # address family
+    metrics = []
+    for family, cache_type in CACHE_TYPES.items():
+        data = read_cache_stats(cache_type)
+        data["family"] = family
+        metrics.append(data)
+    print(json.dumps(metrics))
+
+
+def sensu(args):
+    entries = {}
+    overflows = {}
+    thresholds = {}
+    for family, cache_type in CACHE_TYPES.items():
+        data = read_cache_stats(cache_type)
+        overflows[family] = data["table_fulls"]
+        entries[family] = data["entries"]
+
+        thresholds[family] = read_table_threshold(family)
+
+    with open(args.state_file, "a+b") as fh:
+        fh.seek(0)
+        data = fh.read()
+
+        # only attempt to decode json if the file is non-empty. file
+        # is empty on first creation.
+        state = None
+        if len(data) > 0:
+            state = json.loads(data.decode("utf-8"))
+
+        fh.seek(0)
+        fh.truncate()
```

Comment:
This is an unsafe write: truncate and write can both end up with partial data in the file. This is in itself not a problem if we catch this when loading data. The two options are: make the write safe (write to a tmpfile and rename) or catch errors and start with an empty state. In this situation I'm recommending catching the error and starting fresh.

Reply:
Because this was ambiguous (and everything that can happen will happen): keep the optimistic write path(s) but respond to errors when parsing the json and just start fresh.

Reply:
That's one part that I hate with Python's json decoding: JSON already specifies UTF-8 but the json module operates on (unicode) strings so we have to handle parsing ourselves.

There is a tiny crash situation still left in here: the data might be an invalid UTF-8 byte sequence and that will fail with a different error that we won't recover from. A general except would have been fine here.